include/seqan3/io/detail/misc_output.hpp

```diff
@@ -0,0 +1,103 @@
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \brief Provides various utility functions required only for output.
+ * \author Hannes Hauswedell <hannes.hauswedell AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <iostream>
+#include <string>
+#include <tuple>
+
+#include <seqan3/core/concept/core_language.hpp>
+#ifdef SEQAN3_HAS_BZIP2
+    #include <seqan3/contrib/stream/bz2_ostream.hpp>
+#endif
+#ifdef SEQAN3_HAS_ZLIB
+    #include <seqan3/contrib/stream/gz_ostream.hpp>
+#endif
+
+namespace seqan3::detail
+{
+
+/*!\brief Depending on the given filename/extension, create a compression stream or just forward the primary stream.
+ * \param[in] primary_stream The primary (uncompressed) stream for writing.
+ * \param[in,out] file_name  The associated filename; compression extensions will be stripped.
+ * \returns A pointer to the secondary stream with defaulted or NOP'ed deleter.
+ * \throws seqan3::file_open_error If a compression-extension is used, but is not supported/available.
+ */
+template<char_concept char_t>
+inline auto make_secondary_ostream(std::basic_ostream<char_t> & primary_stream, filesystem::path & file_name)
+    -> std::unique_ptr<std::basic_ostream<char_t>, std::function<void(std::basic_ostream<char_t>*)>>
+{
+    // don't assume ownership
+    constexpr auto stream_deleter_nop      = [] (std::basic_ostream<char_t> *) {};
+    // assume ownership
+    constexpr auto stream_deleter_default  = [] (std::basic_ostream<char_t> * ptr) { delete ptr; };
+
+    std::string extension = file_name.extension().string();
+
+    if ((extension == ".gz") || (extension == ".bgzf") || (extension == ".bam"))
+    {
+        //TODO: always write bgzf even when gz was requested? It's compatible and faster
+    #ifdef SEQAN3_HAS_ZLIB
+        if (extension != ".bam") // remove extension except for bam
+            file_name.replace_extension("");
```

Comment:
After compression/decompression is applied, the format is detected via the "extension before the extension" so we strip the final extension. Except for `.bam` because `.bam` is implicitly `.bam.gz` and removing `.bam` would strip the format.